engine/kv_engine_cleaner.cpp

```diff
@@ -0,0 +1,466 @@
+        pending_purge_dls.emplace_back(
+            PendingPurgeOldRecords{old_dl_records, new_ts});
+        old_dl_records.clear();
+      }
+
+      if (outdated_records.size() > kMaxCachedOldRecords) {
+        pending_purge_outdated.emplace_back(
+            PendingPurgeOutDatedRecords{outdated_records, new_ts});
+        outdated_records.clear();
+      }
+
+      {  // purge pending old string records
+        auto iter = pending_purge_strings.begin();
+        while (iter != pending_purge_strings.end()) {
+          if (iter->release_time < version_controller_.OldestSnapshotTS()) {
+            destroyOldStringRecords(iter->records);
+            iter++;
+          } else {
+            break;
+          }
+        }
+        pending_purge_strings.erase(pending_purge_strings.begin(), iter);
+      }
+
+      {  // purge pending old dl records
+        auto iter = pending_purge_dls.begin();
+        while (iter != pending_purge_dls.end()) {
+          if (iter->release_time < version_controller_.OldestSnapshotTS()) {
+            destroyOldDLRecords(iter->records);
+            iter++;
+          } else {
+            break;
+          }
+        }
+        pending_purge_dls.erase(pending_purge_dls.begin(), iter);
+      }
+
+      {  // purge pending outdated records
+        auto outdated_iter = pending_purge_outdated.begin();
+        while (outdated_iter != pending_purge_outdated.end()) {
+          std::vector<SpaceEntry> outdated_entries;
+          if (outdated_iter->release_time <
+              version_controller_.OldestSnapshotTS()) {
+            purgeOutDatedRecords(outdated_iter->outdated_records,
+                                 &outdated_entries);
+            pending_free_spaces.emplace_back(
+                PendingFreeSpaceEntries{outdated_entries, new_ts});
+            outdated_iter++;
+          } else {
+            break;
+          }
+        }
+        pending_purge_outdated.erase(pending_purge_outdated.begin(),
+                                     outdated_iter);
+      }
+
+      {  // Free pending space entries
+        auto iter = pending_free_spaces.begin();
+        while (iter != pending_free_spaces.end()) {
+          if (iter->release_time < version_controller_.OldestSnapshotTS()) {
+            pmem_allocator_->BatchFree(iter->entries);
+            iter++;
+          } else {
+            break;
+          }
+        }
+        pending_free_spaces.erase(pending_free_spaces.begin(), iter);
+      }
+
+      // Destroy skiplist
+      if (!outdated_skip_lists.empty()) {
+        auto& ts_skiplist = outdated_skip_lists.front();
+        if (ts_skiplist.first < min_snapshot_ts) {
+          ts_skiplist.second->DestroyAll();
+          removeSkiplist(ts_skiplist.second->ID());
+          outdated_skip_lists.pop_front();
+        }
+      }
+
+      // Destroy list
+      if (!outdated_lists.empty()) {
+        auto& ts_list = outdated_lists.front();
```

Comment:
Since this thread is the only thread clearing outdated_lists and the slot is no longer locked, this thread should clean all the outdated List, Hash and Skiplist it can remove every few slots.